neuraxle/steps/loop.py

```diff
@@ -34,57 +34,96 @@ def __init__(
 
-        for di, eo in zip(data_inputs, expected_outputs):
-            self.wrapped = self.wrapped.fit(di, eo)
+        for current_id, di, eo in data_container:
+            self.wrapped, output = self.wrapped.handle_fit(
+                DataContainer(current_ids=None, data_inputs=di, expected_outputs=eo),
+                context
+            )
 
-        return self
+            output_data_container.append(
+                current_id,
+                output.data_inputs,
+                output.expected_outputs
+            )
+
+        current_ids = self.hash(data_container.current_ids, self.get_hyperparams())
+        output_data_container.set_current_ids(current_ids)
 
-    def transform(self, data_inputs):
+        return self, output_data_container
+
+    def handle_transform(self, data_container: DataContainer, context: ExecutionContext):
         """
         Transform each step for each data inputs, and expected outputs
 
-        :param data_inputs:
+        :param data_container: data container
+        :type data_container: DataContainer
+        :param context: execution context
+        :type context: ExecutionContext
         :return: self
         """
-        outputs = []
-        for di in data_inputs:
-            output = self.wrapped.transform(di)
-            outputs.append(output)
+        output_data_container = ListDataContainer.empty()
+
+        for current_id, di, eo in data_container:
+            output = self.wrapped.handle_transform(
+                DataContainer(current_ids=None, data_inputs=di, expected_outputs=eo),
+                context
+            )
 
-        return outputs
+            output_data_container.append(
+                current_id,
+                output.data_inputs,
+                output.expected_outputs
+            )
 
-    def fit_transform(self, data_inputs, expected_outputs=None):
+        current_ids = self.hash(data_container.current_ids, self.get_hyperparams())
+        output_data_container.set_current_ids(current_ids)
+
+        return output_data_container
+
+    def handle_fit_transform(self, data_container: DataContainer, context: ExecutionContext):
         """
         Fit transform each step for each data inputs, and expected outputs
 
-        :param data_inputs:
-        :param expected_outputs:
-        :return: self
+        :param data_container: data container to fit transform
+        :type data_container: DataContainer
+        :param context: execution context
+        :type context: ExecutionContext
+
+        :return: self, transformed_data_container
         """
-        if expected_outputs is None:
-            expected_outputs = [None] * len(data_inputs)
+        output_data_container = ListDataContainer.empty()
+
+        for current_id, di, eo in data_container:
+            self.wrapped, output = self.wrapped.handle_fit_transform(
+                DataContainer(current_ids=None, data_inputs=di, expected_outputs=eo),
```

Comment:
Notice this issue: #172

We want to increase awareness of users when they need to create custom data containers from Nones to avoid doing mistakes. So we want to do issue 172 to make it clean and more rigid.

Reply:
thanks I like this. will be done in a future pr !